pkg/controller/direct/tags/tagbinding_controller.go

```diff
@@ -0,0 +1,314 @@
+			}
+			found = tagBinding
+		}
+	}
+	if found == nil {
+		return false, nil
+	}
+
+	a.actual = found
+	return true, nil
+}
+
+func (a *TagsTagBindingAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	// There no FQN until after creation (server generated id).
+
+	log := klog.FromContext(ctx)
+	log.V(0).Info("creating TagsTagBinding")
+
+	req := &pb.CreateTagBindingRequest{
+		TagBinding: direct.ProtoClone(a.desired),
+	}
+
+	op, err := a.tagBindingsClient.CreateTagBinding(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating TagsTagBinding: %w", err)
+	}
+
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("waiting for creation of TagsTagBinding: %w", err)
+	}
+	log.V(0).Info("created TagsTagBinding", "name", created.GetName())
+
+	// For compatibility, we set spec.resourceID after creation because this is a server-generated-id resource that we are migrating from terraform/DCL.
+	// More info in docs/ai/server-generated-id.md
+	resourceID := strings.TrimPrefix(created.GetName(), "tagBindings/")
+	if err := createOp.SetSpecResourceID(ctx, resourceID); err != nil {
+		return err
+	}
+
+	return a.updateStatus(ctx, createOp, created)
+}
+
+func (a *TagsTagBindingAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	fqn := a.id.String()
+
+	diff, updateMask, err := a.changedFields(ctx)
+	if err != nil {
+		return fmt.Errorf("getting changed fields for TagsTagBinding %q: %w", fqn, err)
+	}
+
+	structuredreporting.ReportDiff(ctx, diff)
+
+	latest := a.actual
+	if len(updateMask.Paths) != 0 {
+		return fmt.Errorf("cannot update TagsTagBinding %q: fields changed: %v; TagBindings are immutable after creation", fqn, updateMask.Paths)
+	}
+
+	return a.updateStatus(ctx, updateOp, latest)
+}
+
+func (a *TagsTagBindingAdapter) updateStatus(ctx context.Context, op directbase.Operation, latest *pb.TagBinding) error {
+	status := &krm.TagsTagBindingStatus{}
+
+	// NOTYET: observedState
+	// {
+	// 	mapCtx := &direct.MapContext{}
+	// 	status.ObservedState = TagsTagBindingObservedState_v1alpha1_FromProto(mapCtx, latest)
+	// 	if mapCtx.Err() != nil {
+	// 		return mapCtx.Err()
+	// 	}
+	// }
+
+	status.ExternalRef = direct.PtrTo(latest.GetName())
+
+	// Legacy status fields
+	status.Name = direct.PtrTo(strings.TrimPrefix(latest.GetName(), "tagBindings/"))
+
+	return op.UpdateStatus(ctx, status, nil)
+}
+
+func (a *TagsTagBindingAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	fqn := a.id.String()
+	if a.actual == nil {
+		return nil, fmt.Errorf("TagsTagBinding %q not found", fqn)
+	}
+
+	obj := &krm.TagsTagBinding{}
+
+	{
+		mapCtx := &direct.MapContext{}
+		spec := TagsTagBindingSpec_FromProto(mapCtx, a.actual)
+		if mapCtx.Err() != nil {
+			return nil, mapCtx.Err()
+		}
+		obj.Spec = *spec
+	}
+
+	obj.SetGroupVersionKind(krm.TagsTagBindingGVK)
+	name := a.actual.TagValueNamespacedName
+	name = strings.ReplaceAll(name, "/", "_")
```

Comment:
COMMENT: Assuming this is reversable because the "namespace" is a camel cased enumeration and the "id" is service generated, so "_" should not occur in either.

Reply:
It's more that we need _a_ name that is semi-unique and a valid k8s object name.